template/upstream.c.erb

```diff
@@ -55,20 +62,46 @@ void btx_register_callbacks_<%= e.name_sanitized %>(
     s->name = "<%= e.name %>";
     s->dispatcher = (void *)&btx_dispatch_<%= e.name_sanitized %>;
     s->callbacks = calloc(1, sizeof(callbacks_t));
-    utarray_new(s->callbacks->event_callbacks, &ut_ptr_icd);
     // 2. Register it
     HASH_ADD_KEYPTR(hh, *name_to_dispatcher, s->name, strlen(s->name), s);
   }
-  // Add the callbacks to the array
-  utarray_push_back(s->callbacks->event_callbacks, &callback);
+
+  <% e.dispatchers.each do |d| %>
+    if(strcmp("<%= d.type %>", callback_type) == 0){
+      // Add the callbacks to the array
+      if (!s->callbacks-><%= d.type %>)
+        utarray_new(s->callbacks-><%= d.type %>, &ut_ptr_icd);
+      utarray_push_back(s->callbacks-><%= d.type %>, &callback);
+    }
+
+  <% end %>
 }
+
+void btx_register_callbacks_<%= e.name_sanitized %>(
+  void *btx_handle, <%= e.name_sanitized %>_callback_f *callback) {
+  _btx_register_callbacks_<%= e.name_sanitized %>(btx_handle, "event", (void *)callback);
+}
+
 <% end %>
+<% matching_dispatchers.each do |m| %>
+void btx_register_matching_callbacks_<%= m.name_sanitized %>(
+  void *btx_handle, <%= m.name_sanitized %>_callback_f *callback){
+  <% event_class_dispatchers.each do |e| %>
+  <% if e.dispatchers.any? { |d| d.name == m.name } %>
```

Comment:
Why you need a `if` here?
why this event class should have a dispatchers that you don't want to call?

Reply:
Same register_callbacks problem. In the matching callback we just generate the `register_event_callback` of the events that match.